### ISSUE TITLE
yolox ONNXRuntime cuda test bug

### ISSUE THREAD

**Issue by NagatoYuki0943** [open-mmlab/mmdeploy#382]
Thanks for your bug report. We appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/faq.md) but cannot get the expected help.
3. The bug has not been fixed in the latest version.

**Describe the bug**

I exported yolox-s model to onnx format. Something went wrong when I tested its speed on cuda, and I found a way to solve it.

**Reproduction**

1. What command or script did you run?

```cmd
python tools/test.py `
    configs/mmdet/detection/detection_onnxruntime_dynamic.py `
    ../mmdetection/configs/yolox/yolox_s_self_voc.py `
    --model work_dir/self/yolox-s-sgd0.001-300-ort-dynamic/end2end.onnx `
    --speed-test `
    --log2file work_dir/self/yolox-s-sgd0.001-300-ort-dynamic/logs-cuda.txt `
    --device cuda:0
```

**Environment**
'''
(mm) PS D:\ai\code\mm\mmdeploy> python .\tools\check_env.py
2022-04-19 16:41:34,720 - mmdeploy - INFO - 

2022-04-19 16:41:34,720 - mmdeploy - INFO - **********Environmental information**********
'tail' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
'head' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
2022-04-19 16:41:36,438 - mmdeploy - INFO - sys.platform: win32
2022-04-19 16:41:36,438 - mmdeploy - INFO - Python: 3.8.13 (default, Mar 28 2022, 06:59:08) [MSC v.1916 64 bit (AMD64)]
2022-04-19 16:41:36,438 - mmdeploy - INFO - CUDA available: True
2022-04-19 16:41:36,438 - mmdeploy - INFO - GPU 0: NVIDIA GeForce MX250
2022-04-19 16:41:36,438 - mmdeploy - INFO - CUDA_HOME: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.3
2022-04-19 16:41:36,438 - mmdeploy - INFO - NVCC: Not Available
2022-04-19 16:41:36,438 - mmdeploy - INFO - GCC: n/a
2022-04-19 16:41:36,438 - mmdeploy - INFO - PyTorch: 1.10.2+cu113
2022-04-19 16:41:36,438 - mmdeploy - INFO - PyTorch compiling details: PyTorch built with:
  - C++ Version: 199711
  - MSVC 192829337
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 2019
  - LAPACK is enabled (usually provided by MKL)
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,co
de=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;ar
ch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.4
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=C:/w/b/windows/tmp_bin/sccache-cl.exe, CXX_FLAGS=/DWIN32
 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental -IC:/w/b/windows/mkl/include -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOCUPTI -DUSE_FBGEMM -DUSE_XN
NPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.2, US
E_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=OFF, USE_OPENMP=ON,    

2022-04-19 16:41:36,438 - mmdeploy - INFO - TorchVision: 0.11.3+cu113
2022-04-19 16:41:36,438 - mmdeploy - INFO - OpenCV: 4.5.5
2022-04-19 16:41:36,438 - mmdeploy - INFO - MMCV: 1.4.8
2022-04-19 16:41:36,438 - mmdeploy - INFO - MMCV Compiler: MSVC 192930140
2022-04-19 16:41:36,438 - mmdeploy - INFO - MMCV CUDA Compiler: 11.3
2022-04-19 16:41:36,438 - mmdeploy - INFO - MMDeploy: 0.4.0+957fd58
2022-04-19 16:41:36,438 - mmdeploy - INFO -

2022-04-19 16:41:36,438 - mmdeploy - INFO - **********Backend information**********
loading mmdeploy_cpu_device ...
loading mmdeploy_cuda_device ...
loading mmdeploy_graph ...
loading mmdeploy_directory_model ...
[2022-04-19 16:41:36.719] [mmdeploy] [info] [model.cpp:95] Register 'DirectoryModel'
loading mmdeploy_transform ...
loading mmdeploy_cpu_transform_impl ...
loading mmdeploy_cuda_transform_impl ...
loading mmdeploy_transform_module ...
loading mmdeploy_trt_net ...
loading mmdeploy_ort_net ...
loading mmdeploy_net_module ...
loading mmdeploy_mmcls ...
loading mmdeploy_mmdet ...
loading mmdeploy_mmseg ...
loading mmdeploy_mmocr ...
loading mmdeploy_mmedit ...
loading mmdeploy_mmpose ...
2022-04-19 16:41:36,809 - mmdeploy - INFO - onnxruntime: 1.11.0 ops_is_avaliable : True
2022-04-19 16:41:36,809 - mmdeploy - INFO - tensorrt: 8.2.3.0   ops_is_avaliable : True
2022-04-19 16:41:36,809 - mmdeploy - INFO - ncnn: None  ops_is_avaliable : False
2022-04-19 16:41:36,809 - mmdeploy - INFO - pplnn_is_avaliable: False
2022-04-19 16:41:36,824 - mmdeploy - INFO - openvino_is_avaliable: False
2022-04-19 16:41:36,824 - mmdeploy - INFO -

2022-04-19 16:41:36,824 - mmdeploy - INFO - **********Codebase information**********
2022-04-19 16:41:36,824 - mmdeploy - INFO - mmdet:      2.20.0
2022-04-19 16:41:36,824 - mmdeploy - INFO - mmseg:      0.21.1
2022-04-19 16:41:36,824 - mmdeploy - INFO - mmcls:      0.19.0
2022-04-19 16:41:36,831 - mmdeploy - INFO - mmocr:      None
2022-04-19 16:41:36,831 - mmdeploy - INFO - mmedit:     None
2022-04-19 16:41:36,831 - mmdeploy - INFO - mmdet3d:    None
2022-04-19 16:41:36,831 - mmdeploy - INFO - mmpose:     None
'''

**Error traceback**

If applicable, paste the error trackback here.

```none
(mm) PS D:\ai\code\mm\mmdeploy> python tools/test.py `
>>     configs/mmdet/detection/detection_onnxruntime_dynamic.py `
>>     ../mmdetection/configs/yolox/yolox_s_self_voc.py `
>>     --model work_dir/self/yolox-s-sgd0.001-300-ort-dynamic/end2end.onnx `
>>     --speed-test `
>>     --log2file work_dir/self/yolox-s-sgd0.001-300-ort-dynamic/logs-cuda.txt `
>>     --device cuda:0
loading mmdeploy_cpu_device ...
loading mmdeploy_cuda_device ...
loading mmdeploy_graph ...
loading mmdeploy_directory_model ...
[2022-04-19 16:34:01.269] [mmdeploy] [info] [model.cpp:95] Register 'DirectoryModel'
loading mmdeploy_transform ...
loading mmdeploy_cpu_transform_impl ...
loading mmdeploy_cuda_transform_impl ...
loading mmdeploy_transform_module ...
loading mmdeploy_trt_net ...
loading mmdeploy_ort_net ...
loading mmdeploy_net_module ...
loading mmdeploy_mmcls ...
loading mmdeploy_mmdet ...
loading mmdeploy_mmseg ...
loading mmdeploy_mmocr ...
loading mmdeploy_mmedit ...
loading mmdeploy_mmpose ...
2022-04-19 16:34:10,159 - mmdeploy - INFO - Successfully loaded onnxruntime custom ops from             d:\ai\code\mm\mmdeploy\build\bin\Release\mmdeploy_onnxruntim
e_ops.dll
[                                                  ] 0/281, elapsed: 0s, ETA:loading mmdeploy_cpu_device ...
loading mmdeploy_cuda_device ...
loading mmdeploy_graph ...
loading mmdeploy_directory_model ...
[2022-04-19 16:34:14.114] [mmdeploy] [info] [model.cpp:95] Register 'DirectoryModel'
loading mmdeploy_transform ...
loading mmdeploy_cpu_transform_impl ...
loading mmdeploy_cuda_transform_impl ...
loading mmdeploy_transform_module ...
loading mmdeploy_trt_net ...
loading mmdeploy_ort_net ...
loading mmdeploy_net_module ...
loading mmdeploy_mmcls ...
loading mmdeploy_mmdet ...
loading mmdeploy_mmseg ...
loading mmdeploy_mmocr ...
loading mmdeploy_mmedit ...
loading mmdeploy_mmpose ...
Traceback (most recent call last):
  File "tools/test.py", line 138, in <module>
    main()
  File "tools/test.py", line 127, in main
    outputs = task_processor.single_gpu_test(model, data_loader,
  File "d:\ai\code\mm\mmdeploy\mmdeploy\codebase\base\task.py", line 137, in single_gpu_test
    return self.codebase_class.single_gpu_test(model, data_loader, show,
  File "d:\ai\code\mm\mmdeploy\mmdeploy\codebase\mmdet\deploy\mmdetection.py", line 142, in single_gpu_test
    outputs = single_gpu_test(model, data_loader, show, out_dir, **kwargs)
  File "D:\Anaconda3\envs\mm\lib\site-packages\mmdet\apis\test.py", line 28, in single_gpu_test
    result = model(return_loss=False, rescale=True, **data)
  File "D:\Anaconda3\envs\mm\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
    return super().forward(*inputs, **kwargs)
  File "D:\Anaconda3\envs\mm\lib\site-packages\torch\nn\parallel\data_parallel.py", line 166, in forward
    return self.module(*inputs[0], **kwargs[0])
  File "D:\Anaconda3\envs\mm\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "d:\ai\code\mm\mmdeploy\mmdeploy\codebase\mmdet\deploy\object_detection_model.py", line 220, in forward
    dets[:, :4] /= scale_factor
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

```

**Bug fix**

If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

Under the line https://github.com/open-mmlab/mmdeploy/blob/master/mmdeploy/codebase/mmdet/deploy/object_detection_model.py#L210 add this code could solve the question.
```python
        ...
        for i in range(batch_size):
            dets, labels = batch_dets[i], batch_labels[i]
            dets = dets.to(device=torch.device(self.device))    # add this line
            if rescale:
        ...
```

run again.
```
(mm) PS D:\ai\code\mm\mmdeploy> python tools/test.py `
>>     configs/mmdet/detection/detection_onnxruntime_dynamic.py `
>>     ../mmdetection/configs/yolox/yolox_s_self_voc.py `
>>     --model work_dir/self/yolox-s-sgd0.001-300-ort-dynamic/end2end.onnx `
>>     --speed-test `
>>     --log2file work_dir/self/yolox-s-sgd0.001-300-ort-dynamic/logs-cuda.txt `
>>     --device cuda:0
loading mmdeploy_cpu_device ...
loading mmdeploy_cuda_device ...
loading mmdeploy_graph ...
loading mmdeploy_directory_model ...
[2022-04-19 16:34:46.463] [mmdeploy] [info] [model.cpp:95] Register 'DirectoryModel'
loading mmdeploy_transform ...
loading mmdeploy_cpu_transform_impl ...
loading mmdeploy_cuda_transform_impl ...
loading mmdeploy_transform_module ...
loading mmdeploy_trt_net ...
loading mmdeploy_ort_net ...
loading mmdeploy_net_module ...
loading mmdeploy_mmcls ...
loading mmdeploy_mmdet ...
loading mmdeploy_mmseg ...
loading mmdeploy_mmocr ...
loading mmdeploy_mmedit ...
loading mmdeploy_mmpose ...
2022-04-19 16:34:52,636 - mmdeploy - INFO - Successfully loaded onnxruntime custom ops from             d:\ai\code\mm\mmdeploy\build\bin\Release\mmdeploy_onnxruntim
e_ops.dll
[                                                  ] 0/281, elapsed: 0s, ETA:loading mmdeploy_cpu_device ...
loading mmdeploy_cuda_device ...
loading mmdeploy_graph ...
loading mmdeploy_directory_model ...
loading mmdeploy_mmpose ...
[>>>>>>>>>>>>>>>>>>>                               ] 109/281, 4.9 task/s, elapsed: 22s, ETA:    35s2022-04-19 16:35:16,965 - test - INFO - [__ort_execute]-110 times per count: 83.93 ms, 11.91 FPS
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>             ] 209/281, 6.1 task/s, elapsed: 34s, ETA:    12s2022-04-19 16:35:28,554 - test - INFO - [__ort_execute]-210 times per count: 95.68 ms, 10.45 FPS
[>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>] 281/281, 6.2 task/s, elapsed: 45s, ETA:     0s

```

**Comment by irexyc** [comment]
Thanks, I will check it.

**Comment by lvhan028** [comment]
@NagatoYuki0943, we'll appreciate it a lot if you can make a PR to MMDeploy.

**Comment by lvhan028** [comment]
@NagatoYuki0943 fixed it in PR #438.

**Comment by chl916185** [comment]
python.exe E:/mmcv_test/mmdeploy/tools/deploy.py
loading mmdeploy_execution ...
failed to load library mmdeploy_execution
loading mmdeploy_cpu_device ...
failed to load library mmdeploy_cpu_device
loading mmdeploy_graph ...
failed to load library mmdeploy_graph
loading mmdeploy_directory_model ...
failed to load library mmdeploy_directory_model
loading mmdeploy_transform ...
failed to load library mmdeploy_transform
loading mmdeploy_cpu_transform_impl ...
failed to load library mmdeploy_cpu_transform_impl
loading mmdeploy_transform_module ...
failed to load library mmdeploy_transform_module
loading mmdeploy_ort_net ...
failed to load library mmdeploy_ort_net
loading mmdeploy_net_module ...
failed to load library mmdeploy_net_module
loading mmdeploy_mmcls ...
failed to load library mmdeploy_mmcls
loading mmdeploy_mmdet ...
failed to load library mmdeploy_mmdet
loading mmdeploy_mmseg ...
failed to load library mmdeploy_mmseg
loading mmdeploy_mmocr ...
failed to load library mmdeploy_mmocr
loading mmdeploy_mmedit ...
failed to load library mmdeploy_mmedit
loading mmdeploy_mmpose ...
failed to load library mmdeploy_mmpose
loading mmdeploy_mmrotate ...
failed to load library mmdeploy_mmrotate
loading mmdeploy_execution ...
failed to load library mmdeploy_execution
loading mmdeploy_cpu_device ...
failed to load library mmdeploy_cpu_device
loading mmdeploy_graph ...
failed to load library mmdeploy_graph
loading mmdeploy_directory_model ...
failed to load library mmdeploy_directory_model
loading mmdeploy_transform ...
failed to load library mmdeploy_transform
loading mmdeploy_cpu_transform_impl ...
failed to load library mmdeploy_cpu_transform_impl
loading mmdeploy_transform_module ...
failed to load library mmdeploy_transform_module
loading mmdeploy_ort_net ...
failed to load library mmdeploy_ort_net
loading mmdeploy_net_module ...
failed to load library mmdeploy_net_module
loading mmdeploy_mmcls ...
failed to load library mmdeploy_mmcls
loading mmdeploy_mmdet ...
failed to load library mmdeploy_mmdet
loading mmdeploy_mmseg ...
failed to load library mmdeploy_mmseg
loading mmdeploy_mmocr ...
failed to load library mmdeploy_mmocr
loading mmdeploy_mmedit ...
failed to load library mmdeploy_mmedit
loading mmdeploy_mmpose ...
failed to load library mmdeploy_mmpose
loading mmdeploy_mmrotate ...
failed to load library mmdeploy_mmrotate
loading mmdeploy_execution ...
failed to load library mmdeploy_execution
loading mmdeploy_cpu_device ...
failed to load library mmdeploy_cpu_device
loading mmdeploy_graph ...
failed to load library mmdeploy_graph
loading mmdeploy_directory_model ...
failed to load library mmdeploy_directory_model
loading mmdeploy_transform ...
failed to load library mmdeploy_transform
loading mmdeploy_cpu_transform_impl ...
failed to load library mmdeploy_cpu_transform_impl
loading mmdeploy_transform_module ...
failed to load library mmdeploy_transform_module
loading mmdeploy_ort_net ...
failed to load library mmdeploy_ort_net
loading mmdeploy_net_module ...
failed to load library mmdeploy_net_module
loading mmdeploy_mmcls ...
failed to load library mmdeploy_mmcls
loading mmdeploy_mmdet ...
failed to load library mmdeploy_mmdet
loading mmdeploy_mmseg ...
failed to load library mmdeploy_mmseg
loading mmdeploy_mmocr ...
failed to load library mmdeploy_mmocr
loading mmdeploy_mmedit ...
failed to load library mmdeploy_mmedit
loading mmdeploy_mmpose ...
failed to load library mmdeploy_mmpose
loading mmdeploy_mmrotate ...
failed to load library mmdeploy_mmrotate
@lvhan028 how to solve it?
